File: merge_data.py
import json
import re

print('🔄 Starting data merge process...\n')

# Read enriched products
print('📖 Reading enriched_products.json...')
with open('enriched_products.json', 'r', encoding='utf-8') as f:
    enriched_products = json.load(f)
print(f'   Found {len(enriched_products)} enriched products\n')

# Read store.ts
print('📖 Reading lib/store.ts...')
with open('lib/store.ts', 'r', encoding='utf-8') as f:
    store_content = f.read()

# Create enrichment map
print('🗺️  Creating enrichment map...')
enrichment_map = {}
for product in enriched_products:
    pid = product.get('id')
    if pid:
        enrichment_map[pid] = {
            'isbn': product.get('isbn'),
            'authors': product.get('authors'),
            'publisher': product.get('publisher'),
            'publicationDate': product.get('publicationDate'),
            'fullDescription': product.get('fullDescription')
        }
print(f'   Created map with {len(enrichment_map)} enriched entries\n')

# Parse products from store.ts
print('🔍 Parsing existing products from store.ts...')
match = re.search(r'export const products: Product\[\] = (\[[\s\S]*?\n\]);', store_content)
if not match:
    print('❌ Could not find products array')
    exit(1)

products_json = match.group(1)
# Fix TypeScript to JSON
products_json = re.sub(r'([{,]\s*)(\w+):', r'\1"\2":', products_json)  # Quote keys
existing_products = json.loads(products_json)
print(f'   Found {len(existing_products)} existing products\n')

# Merge
print('🔀 Merging enriched data...')
enriched_count = 0
for product in existing_products:
    pid = product.get('id')
    if pid and pid in enrichment_map:
        enrichment = enrichment_map[pid]
        for key, value in enrichment.items():
            if value and value != 'N/A':
                product[key] = value
                if key == 'isbn':
                    enriched_count += 1
print(f'   ✅ Enriched {enriched_count} products\n')

# Convert back to TypeScript format
print('📝 Formatting updated products...')
products_ts = json.dumps(existing_products, indent=2, ensure_ascii=False)
products_ts = re.sub(r'"(\w+)":', r'\1:', products_ts)  # Unquote keys

# Replace in store.ts
print('💾 Updating lib/store.ts...')
new_store_content = re.sub(
    r'export const products: Product\[\] = \[[\s\S]*?\n\];',
    f'export const products: Product[] = {products_ts};',
    store_content
)

# Backup
print('📦 Creating backup at lib/store.ts.backup...')
with open('lib/store.ts.backup', 'w', encoding='utf-8') as f:
    f.write(store_content)

# Write
with open('lib/store.ts', 'w', encoding='utf-8') as f:
    f.write(new_store_content)

print('\n✅ SUCCESS! Data merge complete!\n')
print('📊 Summary:')
print(f'   • Total products: {len(existing_products)}')
print(f'   • Enriched products: {enriched_count}')
print(f'   • Backup created: lib/store.ts.backup')
print(f'   • Updated file: lib/store.ts\n')

File: merge_data.js
const fs = require('fs');
const path = require('path');

console.log('🔄 Starting data merge process...\n');

// Read enriched products
const enrichedPath = path.join(__dirname, 'enriched_products.json');
const storePath = path.join(__dirname, 'lib/store.ts');

if (!fs.existsSync(enrichedPath)) {
    console.error('❌ enriched_products.json not found!');
    process.exit(1);
}

console.log('📖 Reading enriched_products.json...');
const enrichedProducts = JSON.parse(fs.readFileSync(enrichedPath, 'utf-8'));
console.log(`   Found ${enrichedProducts.length} enriched products\n`);

console.log('📖 Reading lib/store.ts...');
const storeContent = fs.readFileSync(storePath, 'utf-8');

// Extract the products array from store.ts
const productsMatch = storeContent.match(/export const products: Product\[\] = (\[[\s\S]*?\n\]);/);
if (!productsMatch) {
    console.error('❌ Could not find products array in store.ts');
    process.exit(1);
}

console.log('🔍 Parsing existing products...');
const productsArrayString = productsMatch[1];
const existingProducts = eval(productsArrayString); // Parse the array
console.log(`   Found ${existingProducts.length} existing products\n`);

// Create a map of enriched data by product ID
console.log('🗺️  Creating enrichment map...');
const enrichmentMap = new Map();
enrichedProducts.forEach(product => {
    enrichmentMap.set(product.id, {
        isbn: product.isbn,
        authors: product.authors,
        publisher: product.publisher,
        publicationDate: product.publicationDate,
        editionNumber: product.editionNumber,
        copyrightYear: product.copyrightYear,
        fileSize: product.fileSize,
        fullDescription: product.fullDescription
    });
});
console.log(`   Created map with ${enrichmentMap.size} enriched entries\n`);

// Merge enriched data into existing products
console.log('🔀 Merging enriched data...');
let enrichedCount = 0;
const mergedProducts = existingProducts.map(product => {
    const enrichment = enrichmentMap.get(product.id);
    if (enrichment) {
        enrichedCount++;
        return {
            ...product,
            ...Object.fromEntries(
                Object.entries(enrichment).filter(([_, value]) => value != null && value !== 'N/A')
            )
        };
    }
    return product;
});
console.log(`   ✅ Enriched ${enrichedCount} products\n`);

// Convert merged products back to TypeScript format
console.log('📝 Formatting updated products...');
const formattedProducts = JSON.stringify(mergedProducts, null, 2)
    .replace(/"(\w+)":/g, '$1:') // Remove quotes from keys
    .replace(/: "([^"]+)"/g, (match, value) => {
        // Keep quotes for string values, but format multiline strings nicely
        if (value.includes('\n') || value.length > 200) {
            return `: "${value}"`;
        }
        return match;
    });

// Replace the products array in store.ts
console.log('💾 Updating lib/store.ts...');
const newStoreContent = storeContent.replace(
    /export const products: Product\[\] = \[[\s\S]*?\n\];/,
    `export const products: Product[] = ${formattedProducts};`
);

// Create backup
const backupPath = path.join(__dirname, 'lib/store.ts.backup');
console.log(`📦 Creating backup at lib/store.ts.backup...`);
fs.writeFileSync(backupPath, storeContent);

// Write updated store
fs.writeFileSync(storePath, newStoreContent);

console.log('\n✅ SUCCESS! Data merge complete!\n');
console.log('📊 Summary:');
console.log(`   • Total products: ${existingProducts.length}`);
console.log(`   • Enriched products: ${enrichedCount}`);
console.log(`   • Backup created: lib/store.ts.backup`);
console.log(`   • Updated file: lib/store.ts\n`);

console.log('🎯 Next steps:');
console.log('   1. Restart your dev server (npm run dev)');
console.log('   2. Visit any product page to see the metadata!');
console.log('   3. Check the accordion sections for ISBN, Authors, Publisher, etc.\n');
